iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -24,111 +25,227 @@ namespace Flow {
+  auto rhsIt = rhs.begin();
+  while (lhsIt != lhs.end() && rhsIt != rhs.end()) {
+    if (!pred(*lhsIt++, *rhsIt++)) return false;
+  }
+  if ((lhsIt == lhs.end()) != (rhsIt == rhs.end())) {
+    // Block count mismatch. We do this here so that we avoid the O(n) scan
+    // that would have been required to calculate the size above.
+    return false;
   }
+  return true;
 }
 
-bool areRegionsEquivalent(Region *lhs, Region *rhs) {
-  if (lhs->getBlocks().size() != rhs->getBlocks().size()) {
-    return false;
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs,
+                                       BlockAndValueMapping &parentMapping);
+static bool isStructurallyEquivalentTo(Operation &lhs, Operation &rhs,
+                                       BlockAndValueMapping &parentMapping);
+
+// Recursively compares two regions for structural equivalence.
+// Structural equivalence ensures that operations on both the |lhs| and |rhs|
+// have the same attributes and same use-def structure.
+//
+// Example:
+//   func @lhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//   func @rhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//
+//   assert(isStructurallyEquivalentTo(lhs.getBody(), rhs.getBody()));
+//
+// TODO(#3996): upstream into mlir::OperationEquivalence if this works.
+// TODO(#3996): add symbol ref comparison (add to BlockAndValueMapping).
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs) {
+  BlockAndValueMapping mapping;
+  return isStructurallyEquivalentTo(lhs, rhs, mapping);
+}
+
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs,
+                                       BlockAndValueMapping &parentMapping) {
+  // If the region is isolated we don't want to reuse any parent mapping or
+  // pollute it with our mappings.
+  BlockAndValueMapping scopedMapping;
+  BlockAndValueMapping mapping =
+      lhs.isIsolatedFromAbove() ? scopedMapping : parentMapping;
+
+  // Regions can have arguments; ensure they match in type and add the mapping
+  // from lhs -> rhs so that we can track the usage of them within our
+  // descendants.
+  if (lhs.getNumArguments() != rhs.getNumArguments()) return false;
+  for (auto argPair : llvm::zip(lhs.getArguments(), rhs.getArguments())) {
+    auto &lhsArg = std::get<0>(argPair);
+    auto &rhsArg = std::get<1>(argPair);
+    if (lhsArg.getType() != rhsArg.getType()) return false;
+    mapping.map(lhsArg, rhsArg);
   }
 
-  for (auto blockPair : llvm::zip(lhs->getBlocks(), rhs->getBlocks())) {
+  // Walk each block list together. If the sizes differ we'll handle that after
+  // Build the block mapping so that when we recursive into ops we are able to
+  // map lhs->rhs blocks regardless of their order in the region.
+  //
+  // We walk each block list simultaneously and handle length differences after
```

Comment:
I find this last paragraph is just confusing. I think the whole comment could be reduced to "use compare_ranges to walk the block list in parallel and get a boolean in the case of size mismatch without an O(N) linked-list size query"